Validate origin format based on issue tracker

srafi1 opened the pull request:
Solves #25

srafi1 commented:
I'm not sure how to validate the origin for Jira or Redmine, but this should validate the others properly

preslavmihaylov commented on config/config.go:
```diff
@@ -57,6 +64,33 @@ func NewLocal(cfgPath, basepath string) (*Local, error) {
+	case IssueTrackerRedmine:
+		validOrigin = redmineOriginPattern.MatchString(cfg.Origin)
+		break
+	case IssueTrackerInvalid:
```
Could you convert this switch-case into a map - `map[IssueTracker]*Regexp`.

Afterwards, validation can look like:
```
pattern, !ok := originPatterns[cfg.IssueTracker]
if !ok {
   // return err
}

// validate pattern matches
```

preslavmihaylov replied:
Also, please make sure this map is close to the enum declaration to make things easier for future maintainers once a new integration is added.

srafi1 replied:
resolved in eaf4e4a

preslavmihaylov commented on config/config.go:
```diff
@@ -57,6 +64,33 @@ func NewLocal(cfgPath, basepath string) (*Local, error) {
 		return nil, fmt.Errorf("failed to unmarshal local configuration (%s): %w", cfgPath, err)
 	}
 
+	// Check that origin matches issue tracker's format
```
Could you pull latest master & move this chunk into the `Validate` function for consistency? 

See #46

srafi1 replied:
resolved in 78a7217

preslavmihaylov commented on config/config.go:
```diff
@@ -27,6 +27,13 @@ const (
+var gitlabOriginPattern = regexp.MustCompile(`^(https://)?(www\.)?gitlab\.com/[a-zA-Z0-9\-_]+/[a-zA-Z0-9\-_]+`)
+var pivotalOriginPattern = regexp.MustCompile(`^(https://)?(www\.)?pivotaltracker/n/projects/[0-9]+`)
+// TODO: add redmine issue tracker pattern
+var redmineOriginPattern = regexp.MustCompile(``)
```
For redmine, the regex should match any domain as it is self-hosted (e.g. `https://my-redmine-instance.com`)

There is no expected path appended to the hosts as redmine issues are accessed as `https://my-redmine-instance.com/issues/3`

preslavmihaylov replied:
You can test integration \w Redmine by signing up to their demo hosted website - http://demo.redmine.org/

srafi1 replied:
resolved in accb53f

preslavmihaylov commented on config/config.go:
```diff
@@ -27,6 +27,13 @@ const (
 
 var windowsAbsolutePathPattern = regexp.MustCompile("^[A-Z]{1}:")
+// TODO: add jira issue tracker pattern
+var jiraOriginPattern = regexp.MustCompile(``)
```
For Jira, the origin is similar to redmine. It can be any domain (e.g. `https://myjira.com`).

srafi1 replied:
resolved in accb53f

preslavmihaylov reviewed (changes requested):
Thanks for taking the time to put together this PR!

Left some comments & suggestions for validating Jira and redmine.

I can help with testing the new validations for all supported issue trackers once you're ready.

srafi1 commented:
Thanks for the comments. I fixed everything you mentioned, but I have a couple of questions. Why does `Validate()` return a list of errors when it can only ever have one according to the current logic? Also, when I ran the tests with `go test ./testing -v`, many of the tests failed. I ended up running my own tests using `go test ./testing -run Origins -v` and although they pass, I'm not sure if I'm running tests properly if the other ones are failing on my computer.

preslavmihaylov commented on config/config.go:
```diff
@@ -36,6 +36,14 @@ var validIssueTrackers = []IssueTracker{
+	IssueTrackerGithub:  regexp.MustCompile(`^(https://)?(www\.)?github\.com/[a-zA-Z0-9\-_]+/[a-zA-Z0-9\-_]+`),
+	IssueTrackerGitlab:  regexp.MustCompile(`^(https://)?(www\.)?gitlab\.com/[a-zA-Z0-9\-_]+/[a-zA-Z0-9\-_]+`),
+	IssueTrackerPivotal: regexp.MustCompile(`^(https://)?(www\.)?pivotaltracker\.com/n/projects/[0-9]+`),
+	IssueTrackerRedmine: regexp.MustCompile(`[a-zA-Z0-9\-]+\.[a-zA-Z]+$`),
```
This pattern excludes valid origins (i.e subdomains) such as `https://redmine.myorg.com`.

Please extend the pattern to allow such origins & add a test case for that.

srafi1 replied:
fixed in f4977d0 with test case in defa509

preslavmihaylov commented on config/config.go:
```diff
@@ -36,6 +36,14 @@ var validIssueTrackers = []IssueTracker{
 var windowsAbsolutePathPattern = regexp.MustCompile("^[A-Z]{1}:")
 
+var originPatterns = map[IssueTracker]*regexp.Regexp{
+	IssueTrackerJira:    regexp.MustCompile(`[a-zA-Z0-9\-]+\.[a-zA-Z]+$`),
```
This pattern excludes valid origins (i.e subdomains) such as `https://myjira.myorg.com`.

Please extend the pattern to allow such origins & add a test case for that.

srafi1 replied:
fixed in f4977d0 with test case in defa509

preslavmihaylov commented on config/config.go:
```diff
@@ -36,6 +36,14 @@ var validIssueTrackers = []IssueTracker{
+var originPatterns = map[IssueTracker]*regexp.Regexp{
+	IssueTrackerJira:    regexp.MustCompile(`[a-zA-Z0-9\-]+\.[a-zA-Z]+$`),
+	IssueTrackerGithub:  regexp.MustCompile(`^(https://)?(www\.)?github\.com/[a-zA-Z0-9\-_]+/[a-zA-Z0-9\-_]+`),
+	IssueTrackerGitlab:  regexp.MustCompile(`^(https://)?(www\.)?gitlab\.com/[a-zA-Z0-9\-_]+/[a-zA-Z0-9\-_]+`),
```
Some companies have a private instance of gitlab setup. E.g. `https://mygitlab.org.com`.

`todocheck` has support for private gitlab instances but this validation excludes them.

Example - #40 

This pattern should (similarly to jira & redmine) match any (sub)domain.

srafi1 replied:
fixed in 56054bf and 4b08267 with test case in defa509

preslavmihaylov commented on testing/test_configs/valid_github_origin.yaml:
```diff
@@ -0,0 +1,2 @@
+origin: https://github.com/user/project
```
Could you also add valid configs for all supported issue trackers:
 * Without a scheme provided - e.g. `github.com/user/project`
 * With `www` provided - e.g. `www.github.com/user/project` and `https://www.github.com/user/project`

srafi1 replied:
added in defa509

preslavmihaylov commented on testing/todocheck_test.go:
```diff
@@ -330,6 +331,50 @@ func TestInvalidIssueTracker(t *testing.T) {
+		"./test_configs/valid_github_origin.yaml",
+		"./test_configs/valid_gitlab_origin.yaml",
+		"./test_configs/valid_pivotal_origin.yaml",
+		"./test_configs/valid_redmine_origin.yaml",
```
This is a pretty neat way to test all these.
It avoids adding a separate test case for all issue trackers.

Thanks!

preslavmihaylov reviewed (changes requested):
I have a few more comments left to make this fully compatible with what's currently supported & for adding more extensive test cases.

preslavmihaylov commented:
> Thanks for the comments. I fixed everything you mentioned, but I have a couple of questions. Why does `Validate()` return a list of errors when it can only ever have one according to the current logic? Also, when I ran the tests with `go test ./testing -v`, many of the tests failed. I ended up running my own tests using `go test ./testing -run Origins -v` and although they pass, I'm not sure if I'm running tests properly if the other ones are failing on my computer.

The tests all pass when I run them from latest master:
![image](https://user-images.githubusercontent.com/7642330/92203389-bc947e80-ee89-11ea-955d-f5c43acc5e80.png)

Could you try running the test suite from latest master & see if they pass?

As for the `Validate` function, the goal is to batch report all issues with the configuration. 

Currently, your validation is incompatible with the rest but as more validations are added, there will be independent issues \w configurations which can be reported to the user.

preslavmihaylov commented on config/config.go:
```diff
@@ -82,6 +90,14 @@ func (l *Local) Validate() []error {
 		errors = append(errors, err)
 	}
 
+	if 0 == len(errors) {
```
This check is dependent on the order of validations. This shouldn't happen as this piece of validation might as well become first inline in the future.

Instead, change your condition to be independent from the `errors` array by validating that the issue tracker is not empty:
```
if l.IssueTracker != "" {
   ...
}
```

srafi1 replied:
fixed in f4977d0

preslavmihaylov commented:
Blocked by #54

preslavmihaylov commented:
Hey, the PR blocking this one is now merged. Please pull latest master & continue from there (to enable the CI checks).

srafi1 commented:
I figured out the testing issue on my end, added a bunch of tests for the regex, and fixed the regex according to your comments. LMK if there's anything else to add.

preslavmihaylov commented on config/config.go:
```diff
@@ -145,6 +153,13 @@ func (l *Local) Validate() []error {
+	if l.IssueTracker != "" {
+		pattern, ok := originPatterns[l.IssueTracker]
+		if !ok || !pattern.MatchString(l.Origin) {
+			errors = append(errors, fmt.Errorf("origin is not valid for issue tracker: %s", l.IssueTracker))
```
Could you make the err message slightly more informative by writing:
```
fmt.Sprintf("%s is not a valid origin for issue tracker %s", l.Origin, l.IssueTracker)
```

srafi1 replied:
resolved in 16833af

preslavmihaylov reviewed (changes requested):
Everything looks good this time around in terms of implementation, thanks!

I only have a single comment for the err message format used to report issues with invalid origins.
